mlflow/utils/rest_utils.py

```diff
@@ -14,26 +14,76 @@
 from mlflow.utils.string_utils import strip_suffix
 from mlflow.exceptions import MlflowException, RestException
 
-_REST_API_PATH_PREFIX = "/api/2.0"
 RESOURCE_DOES_NOT_EXIST = "RESOURCE_DOES_NOT_EXIST"
+_REST_API_PATH_PREFIX = "/api/2.0"
+_DEFAULT_HEADERS = {"User-Agent": "mlflow-python-client/%s" % __version__}
+# Response codes that generally indicate transient network failures and merit client retries,
+# based on guidance from cloud service providers
+# (https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines)
+_TRANSIENT_FAILURE_RESPONSE_CODES = [
+    408,  # Request Timeout
+    429,  # Too Many Requests
+    500,  # Internal Server Error
+    502,  # Bad Gateway
+    503,  # Service Unavailable
+    504,  # Gateway Timeout
+]
 
-_logger = logging.getLogger(__name__)
 
-_DEFAULT_HEADERS = {"User-Agent": "mlflow-python-client/%s" % __version__}
+def _get_http_response_with_retries(max_retries, backoff_factor, retry_codes, **kwargs):
+    """
+    Performs an HTTP request using Python's `requests` module with an automatic retry policy.
+
+    :max_retries: Maximum total number of retries.
+    :backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP request
+      will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the exp-backoff.
+    :retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :kwargs: Keyword arguments to pass to `requests.Session.request()`
+    """
+    assert 0 <= max_retries < 10
+    assert 0 <= backoff_factor < 120
+    retry = Retry(
+        total=max_retries,
+        connect=max_retries,
+        read=max_retries,
+        redirect=max_retries,
+        status=max_retries,
+        status_forcelist=retry_codes,
+        backoff_factor=backoff_factor,
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    with requests.Session() as http:
+        http.mount("https://", adapter)
+        http.mount("http://", adapter)
+        method = kwargs.pop("method")
+        url = kwargs.pop("url")
+        response = http.request(method, url, **kwargs)
+        return response
 
 
 def http_request(
-    host_creds, endpoint, retries=3, retry_interval=3, max_rate_limit_interval=60, **kwargs
+    host_creds,
+    endpoint,
+    max_retries=3,
+    backoff_factor=5,
```

Comment:
With 3 max retries, and backoff_factor of 5. We expect to retry after failure at 5, 10, 20s. This should be sufficient for most internet conditions.

Reply:
Using `max_retries=3` and `backoff_factor=5` reduces our retry duration for rate limiting from 60 seconds to ~ 35 seconds. This seems acceptable to me, but it's worth calling out.

Reply:
5 seconds is a long time to wait for the first failure. Perhaps we can use a smaller backoff factor with more retries?

Reply:
Changed factor to 2 with 5 retries, total retry time is around 2^6-2 = 62 seconds.